smart_open/tests/test_ftp.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+
+import logging
+import unittest
+from unittest.mock import patch, call
+
+import smart_open
+from smart_open import ftp
+
+
+class FTPOpen(unittest.TestCase):
+    @patch("smart_open.ftp.FTP")
+    def test_open_just_user(self, ftp_obj):
+        ftp_obj().transfercmd().makefile().name = "some/path"
+        smart_open.open("ftp://user@localhost/dir1/dir2/dir3/file")
+        ftp_obj().connect.assert_called_with("localhost", ftp.DEFAULT_PORT)
+        ftp_obj().sendcmd.assert_has_calls([call("USER user")])
+        ftp_obj().transfercmd.assert_called_with("RETR /dir1/dir2/dir3/file")
+
+    @patch("smart_open.ftp.FTP")
+    def test_open_user_pass(self, ftp_obj):
+        ftp_obj().transfercmd().makefile().name = "some/path"
+        smart_open.open("ftp://user:pass@localhost/dir1/dir2/dir3/file")
+        ftp_obj().connect.assert_called_with("localhost", ftp.DEFAULT_PORT)
+        ftp_obj().sendcmd.assert_has_calls([call("USER user"), call("PASS pass")])
+        ftp_obj().transfercmd.assert_called_with("RETR /dir1/dir2/dir3/file")
+
+    @patch("smart_open.ftp.FTP")
+    def test_open_user_pass_port(self, ftp_obj):
+        ftp_obj().transfercmd().makefile().name = "some/path"
+        smart_open.open("ftp://user:pass@localhost:29/dir1/dir2/dir3/file")
+        ftp_obj().connect.assert_called_with("localhost", 29)
+        ftp_obj().sendcmd.assert_has_calls([call("USER user"), call("PASS pass")])
+        ftp_obj().transfercmd.assert_called_with("RETR /dir1/dir2/dir3/file")
+
+    @patch("smart_open.ftp.FTP")
+    def test_open_with_transport_params(self, ftp_obj):
+        ftp_obj().transfercmd().makefile().name = "some/path"
+        smart_open.open(
+            "ftp://user@localhost/dir1/dir2/dir3/file",
+            transport_params={
+                "keyfile": "some_key_file",
```

Comment:
Same outdated note as above for these tests

Reply:
Thank you for bringing this up. I have updated the tests to reflect the changes.